cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java

```diff
@@ -95,14 +96,35 @@ protected final HttpServiceRequest wrapRequest(HttpRequest request) {
    * object. This method is called from handler class generated by {@link HttpHandlerGenerator}.
    */
   @SuppressWarnings("unused")
-  protected final DelayedHttpServiceResponder wrapResponder(HttpResponder responder) {
+  protected final DelayedHttpServiceResponder wrapResponder(HttpResponder responder,
+                                                            final TransactionContext txContext) {
+
+
     MetricsContext collector = this.metricsContext;
     HttpServiceContext serviceContext = context.getServiceContext();
-    if (serviceContext != null && serviceContext.getSpecification() != null) {
+    Preconditions.checkState(serviceContext instanceof TransactionalHttpServiceContext,
```

Comment:
Not with the current implementation. It was due to some old code logic so that `DelegatorContext.getServiceContext` returns a `HttpServiceContext` instead of `TransactionalHttpServiceContext`, hence with the check to guard against potential bug. Will do a refactoring in a separate PR to change `DelegatorContext` to always return `TransactionalHttpServiceContext`.